### ISSUE TITLE
Bad interaction b/w two phase checker and triggers?

### ISSUE THREAD

**Issue by jaybosamiya** [FStarLang/FStar#1396]
```
module Triggers

open FStar.Seq

#set-options "--use_two_phase_tc true"

unfold let (.[]) = index

assume val (====) : #t:Type -> t -> t -> bool

let foo123 (a:seq int) : unit =
  assume (forall x. {:pattern (a.[x] ==== a.[x+1])}
            0 <= x /\ x < length a - 1 ==> a.[x] ==== a.[x+1]);
  assume (length a > 5);
  assert (a.[0] ==== a.[0+1]);
  ()
```

If I either remove the `#set-options "..."` line, or remove the `{:pattern ...}` from the forall, the proof goes through

**Comment by aseemr** [comment]
Hi @jaybosamiya, I started looking at this example. But even without two-phases, it is a bit brittle. For example, this also fails:

```
module Test

open FStar.Seq

#set-options "--max_fuel 0 --max_ifuel 0"

assume val (====) : int -> int -> bool

let foo123 (a:seq int) : unit =
  assume (forall (x:nat). {:pattern (Seq.index a x ==== Seq.index a (x+1))}
            0 <= x /\ x < length a - 1 ==> Seq.index a x ==== Seq.index a (x+1));
  assume (length a > 5);
  assert (Seq.index a 0 ==== Seq.index a (0+1));
  ()
```

I think the reason for brittleness is that, the pattern on the quantifier is `Seq.index a x ==== Seq.index a (x + 1)`, whereas in the query, since `0 + 1` is a constant, it is simply reduced to `Seq.index a 0 == Seq.index a 1`, which means the trigger may not fire (there is no `+` in the query anymore).

On the other hand, this succeeds:

```
module Test

open FStar.Seq

#set-options "--max_fuel 0 --max_ifuel 0"

assume val (====) : int -> int -> bool

let foo123 (a:seq int) : unit =
  assume (forall (x:nat). {:pattern (Seq.index a x ==== Seq.index a (x+1))}
            0 <= x /\ x < length a - 1 ==> Seq.index a x ==== Seq.index a (x+1));
  assume (length a > 5);
  assert (forall (x:nat). Seq.index a 0 ==== Seq.index a (0+1));  //only added a forall x, which is unused
  ()
```

The difference is an additional `forall x` in the final `assert` where `x` is not even used after that (and this succeeds even in two phases). I am not sure about the reason for this brittleness (@nikswamy, @wintersteiger).

You had other examples in your file for two-phases and triggers, can you post some other instance of this? Thanks!

**Comment by jaybosamiya** [comment]
Thanks @aseemr and @nikswamy. It does indeed seem like the `+` in the pattern might be causing this. I took a look over all the other examples for the bad interactions that are there, and all of them seem to involve either a `+` or a `-` in them.

It does seem odd to me that we should face such behavior. I recall someone mentioning that there are 2 possible encodings for theory symbols like `+`, and maybe that's what's causing this?

Maybe we could have a warning for when theory symbols are used in patterns? I know that I've been bitten in the past due to them (see https://github.com/jaybosamiya/poly1305-equivalence/blob/master/UsefulLemmas.fst#L48 for a workaround I had to use), and as far as I remember, recently @R1kM faced similar issues as well.

For now, the following seems to work:
```
module TriggersMod

open FStar.Seq

#set-options "--use_two_phase_tc true"

unfold let (.[]) = index

assume val (====) : #t:Type -> t -> t -> bool

let foo123 (a:seq int) : unit =
  assume (forall x y. {:pattern (a.[x] ==== a.[y])}
            0 <= x /\ x < length a - 1 /\ y = x+1 ==> a.[x] ==== a.[y]);
  assume (length a > 5);
  assert (a.[0] ==== a.[0+1]);
  ()
```

Basically, I added an extra `y = x + 1` in the implication (and replacing the `x+1` with `y` during usage), and the `forall` is over both `x` and `y` now. This way, our pattern stays free of the theory symbol.

Do you see any potential downsides to this fix? Or is it the recommended way of getting around not using theory symbols in the pattern?

Thanks again! :)

**Comment by nikswamy** [comment]
We could issue a warning---there are some legitimate uses, but those are rare.
Note, `=` and its variants (`==`, `===`, etc.) are also theory symbols.
You seem to be working around something by introducing `====` (!). I don't think it's a good idea to use that in a pattern either.

I would probably have written this as:

```
module TriggersMod

open FStar.Seq

#set-options "--use_two_phase_tc true"

unfold let (.[]) = index

let foo123 (a:seq int) : unit =
  assume (forall x. {:pattern (a.[x])}
            0 <= x /\ x < length a - 1 ==> a.[x] = a.[x+1]);
  assume (length a > 5);
  assert (a.[0] = a.[0+1])
```

No need to introduce a double quantifier when y can be computed from x.

**Comment by jaybosamiya** [comment]
Well, `====` was just a random new operator (should've picked a different operator to prevent confusion with equality) that I defined for this particular short PoC. I _don't_ use that to work around anything for equality (`=` or `==` or `===`), and in my actual use case it would be something like `-|>` or `<|-` or ..., but thanks for the heads up :)

In this specific case, picking this shorter quantifier pattern can cause a trigger loop, right? I don't see it causing a trigger loop here, but if I understand correctly, it does seem to have the potential to.

**Comment by nikswamy** [comment]
You're right, that could introduce a matching loop, in general.

But, I also expect that in a more realistic situation you'll have something better to trigger on. For instance:

```
module TriggersMod

open FStar.Seq

#set-options "--use_two_phase_tc true"

unfold let (.[]) = index

let foo123 (a:seq int{length a > 1}) : unit =
  assert (forall x.{:pattern (tail a).[x]} 0 <= x /\ x < length (tail a) ==>
                         (tail a).[x] = a.[x + 1]);
  assert (a.[0 + 1] = (tail a).[0])
```

**Comment by aseemr** [comment]
Closing it for now.
